force-app/main/default/classes/BDI_MatchDonations.cls

```diff
@@ -430,94 +430,105 @@ public with sharing class BDI_MatchDonations implements BDI_IMatchDonations {
-        // the set of Payment fields for our soql
-        set<String> setPmtFields = new set<String>();
-        setPmtFields.add('Id');
-        setPmtFields.add('Name');
-        setPmtFields.add('npe01__Paid__c');
-        setPmtFields.add('npe01__Opportunity__c');
-        setPmtFields.add('npe01__Payment_Amount__c');
-        setPmtFields.add('npe01__Payment_Date__c');
-        setPmtFields.add('npe01__Scheduled_Date__c');
-        setPmtFields.add('npe01__Written_Off__c');
+        BDI_PerfLogger.PerfLog pl = BDI_DataImportService.perfLogger.newPerfLog('SOQL existing Opps', BDI_PerfLogger.QUERY);
+        List<Opportunity> listOpp = Database.query(buildPotentialDonationMatchingQuery(
+            opportunityIds, accountIds, contactIds));
+        pl.stop();
+        return listOpp;           
+    }
 
-        // need to include any Pmt fields we may update, so we can compare against them
-        setPmtFields.addAll(mapDIFieldToPmtField.values());
+    private String buildPotentialDonationMatchingQuery(Set<Id> opportunityIds, Set<Id> accountIds, Set<Id> contactIds) {
+        Integer cFilter = 0;
+        String strSoql = 'SELECT ' + String.join(new List<String>(opportunityFields()), ',');
 
-        // now query for all Open Opportunities from these Contacts or Accounts
-        integer cFilter = 0;
-        string strSoql = 'SELECT ' + string.join(new list<String>(setOppFields), ',');
-        
         // add Payment subquery
-        strSoql += ', (SELECT ' + string.join(new list<String>(setPmtFields), ',');
+        strSoql += ', (SELECT ' + String.join(new List<String>(paymentFields()), ',');
         strSoql += ' FROM npe01__OppPayment__r ' +
-                'WHERE npe01__Paid__c = false ' +
-                'AND npe01__Written_Off__c = false)';
+            'WHERE npe01__Paid__c = false ' +
+            'AND npe01__Written_Off__c = false)';
 
         strSoql += ' FROM Opportunity';
         strSoql += ' WHERE IsClosed = false AND (';
-        if (setOppId.size() > 0) {
+        if (opportunityIds.size() > 0) {
             strSoql += ' Id IN :setOppId ';
             cFilter++;
         }
-        if (setAccId.size() > 0) {
+        if (accountIds.size() > 0) {
             if (cFilter > 0) {
                 strSoql += ' or ';
             }
             strSoql += ' AccountId IN :setAccId ';
             cFilter++;
         }
-        if (setConId.size() > 0) {
+        if (contactIds.size() > 0) {
             if (cFilter > 0) {
                 strSoql += ' OR ';
             }
             strSoql += ' ' + UTIL_Namespace.StrTokenNSPrefix('Primary_Contact__c') + ' IN :setConId ';
             cFilter++;
         }
-        strSoql += ') ';
-        BDI_PerfLogger.PerfLog pl = BDI_DataImportService.perfLogger.newPerfLog('SOQL existing Opps', BDI_PerfLogger.QUERY);
-        list<Opportunity> listOpp = database.query(strSoql);
-        pl.stop();
-        return listOpp;           
+        strSoql += ') LIMIT 50000';
```

Comment:
What do you think about making this method @TestVisible and asserting that the generated query contains `LIMIT 50000`?